### ISSUE TITLE
insert_into_table fails

### ISSUE THREAD

**Issue by aersam** [pacman82/arrow-odbc-py#39]
Hi there!

I'm faking an issue when using `insert_into_table`. The only error message I get is this one: `Failure to execute the sql statement, sending the data to the database`. I use MS Azure SQL Server as target

I use insert_into_table for a number of tables and it works for some, for others it does not. therefore I think it might be related to not supported datatypes or wrong mappings. 

Is there a way to get a more detailed error message?

Data Types I use that might be exotic:
`timestamp[us]` -> `datetime2`
`bool` -> `bit`
`float` -> `double` 

Regards,
Adrian

**Comment by aersam** [comment]
Just to add: this is not urgent (yet ;) ) for me - my first priority is my holiday starting tomorrow ;)

**Comment by aersam** [comment]
Would a duplicate key or nulls in a not-null column cause such an error?

**Comment by pacman82** [comment]
Hello @aersam ,

I see two distinct issues within your original request:

1. Find out why insert fails for your tables from your data (of course several issues could be hiding within that one).
2. Make more ODBC diagnostic records available for Python users.

Of course if `2` is done we might be better equipped to handle `1`. On the other hand we just might find this was the only error ever surfaced by ODBC. Anyhow I created #40 to track that separatly. I understand that problem and given time, I confident I can solve it without further input. Although the implementation might not be trivial.

With `1` the quickest way to victory might be a minimal reproducing example. Let's pick one of your tables which does not work. And see if we can reproduce the behaviour without azure and just an MSSQL database. Then let's reduce the number of columns and so on...

I fear without a small reproducing test this is not actionable for me. If this issue becomes urgent for you I would advice not to wait for `2`. I totally see that this is usefull and might help you to find the reproducing example a lot faster, but I am really unsure how much open source I'll be able to do over the next weeks.

Best, Markus

Ps. If you know Rust, you could of course just use the `arrow-odbc` Rust crate with logging to insert your data. A quick prototype should might tell you if azure does give any helpful information at all, or if this is just something we hope it would do.

**Comment by pacman82** [comment]
`log_to_stderr` has been added in `arrow-odbc 1.1.1`. Calling this at the beginnig of your application or script will result in ODBC diagnostic records being written to standard error. Fancier interactions with the Python logger are possible, but likely there is a price to pay in terms of performance. Let's see if this is enough.

**Comment by aersam** [comment]
log_to_stderr has helped a lot already. Most errors happened because of duplicates. However I have a case where I do get a number of unexplainable empty strings in my target table. I think my source is correct; is there any reason arrow-odbc could insert an empty string instead of the correct value? Not entirely sure if it's caused by arrow-odbc but it seems a bit like it

**Comment by pacman82** [comment]
>  I think my source is correct; is there any reason arrow-odbc could insert an empty string instead of the correct value? Not entirely sure if it's caused by arrow-odbc but it seems a bit like it

I would be better able to help if you provide information like:

What is the source value in the array? What is its arrow type? What is the type of the database column it is inserted to?

**Comment by aersam** [comment]
Of course! I could also try to create a minimal repo. The source is a large_string, the target is nvarchar(38) (don't ask). The source value is `840003056` (and others). it's really a bit strange, it works in 99% but not in all

**Comment by pacman82** [comment]
```python
def test_insert_large_string():
    """
    Insert an arrow table whose schema contains a "large string".
    """
    # Given
    table = "InsertLargeString"
    os.system(f'odbcsv fetch -c "{MSSQL}" -q "DROP TABLE IF EXISTS {table};"')
    os.system(f'odbcsv fetch -c "{MSSQL}" -q "CREATE TABLE {table} (a Varchar(50))"')
    schema = pa.schema([("a", pa.large_string())])

    def iter_record_batches():
        # The string value is not actually large, just the schema information allows it to be
        yield pa.RecordBatch.from_arrays([pa.array(["Hello"])], schema=schema)

    reader = pa.RecordBatchReader.from_batches(schema, iter_record_batches())

    # When
    insert_into_table(
        connection_string=MSSQL, chunk_size=20, table=table, reader=reader
    )

    # Then
    actual = check_output(
        ["odbcsv", "fetch", "-c", MSSQL, "-q", f"SELECT a FROM {table}"]
    )
    assert "a\nHello\n" == actual.decode("utf8")
```

**Comment by pacman82** [comment]
Hello @aersam ,

any new insight into what exactly the symptom is. Either through tests, minimal example or otherwise?

Best, Markus

**Comment by aersam** [comment]
Hi Markus

Sorry, was a bit under time pressure and did a workaround using pyodbc. That was slow, but works. Now I did decide to learn some rust and did a prototype that uses MS SQL Server's bulk_insert which should be the fastest option there is. I'll open source that project in the next days. It's still very limited, but seems to be very fast. And as I learned some rust now I hope I can debug the error myself, just need to find some time

Regards,
Adrian
